plugin/caffe/caffe_data_iter.cc

```diff
@@ -0,0 +1,217 @@
+/*!
+ * Copyright (c) 2015 by Contributors
+ * \file caffe_data_iter.cc
+ * \brief register mnist iterator
+*/
+#include <caffe/proto/caffe.pb.h>
+#include <dmlc/logging.h>
+#include <dmlc/parameter.h>
+#include <mxnet/operator.h>
+#include <atomic>
+
+#include "../../src/operator/operator_common.h"
+#include "caffe_common.h"
+#include "caffe_stream.h"
+#include "caffe_fieldentry.h"
+#include "caffe_blob.h"
+#include "../../src/io/inst_vector.h"
+#include "../../src/io/iter_prefetcher.h"
+#include <../../src/operator/cast-inl.h>
+
+namespace mxnet {
+namespace io {
+
+struct CaffeDataParam : public dmlc::Parameter<CaffeDataParam> {
+  /*! \brief protobuf text */
+  ::caffe::LayerParameter prototxt;
+  /*! \brief maximum iteration count */
+  int max_iterations;  
+  /*! \brief data mode */
+  bool flat;
+
+  DMLC_DECLARE_PARAMETER(CaffeDataParam) {
+    DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
+      .describe("Caffe's layer parameter");
+    DMLC_DECLARE_FIELD(flat).set_default(false)
+      .describe("Augmentation Param: Whether to flat the data into 1D.");
+    DMLC_DECLARE_FIELD(max_iterations).set_lower_bound(1).set_default(10000)
+      .describe("Maximum number of iterations per epoch.");
+  }
+};
+
+struct CaffePrefetchParam : dmlc::Parameter<CaffePrefetchParam> {
+    /*! \brief data type */
+    int dtype;
+    
+    DMLC_DECLARE_PARAMETER(CaffePrefetchParam) {
+      DMLC_DECLARE_FIELD(dtype)
+        .add_enum("float32", mshadow::kFloat32)
+        .add_enum("float64", mshadow::kFloat64)
+        .add_enum("float16", mshadow::kFloat16)
+        .add_enum("uint8", mshadow::kUint8)
+        .add_enum("int32", mshadow::kInt32)
+        .set_default(mshadow::default_type_flag)
+        .describe("Data type.");
+    }
+};
+
+template<typename Dtype>
+class CaffeDataIter : public IIterator<TBlobBatch> {
+
+public:
+  CaffeDataIter(void) : loc_(0), batch_size_(0), channels_(0), width_(0), height_(0) {}
+  virtual ~CaffeDataIter(void) {}
+    
+  // intialize iterator loads data in
+  virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+    std::map<std::string, std::string> kmap(kwargs.begin(), kwargs.end());
+    param_.InitAllowUnknown(kmap);
+
+    // Caffe seems to understand phase inside an "include {}" block
```

Comment:
Later in caffe, top-level phase member of prototxt is checked. This causes both data layers to be treated as "train". I did not see how normal caffe gets around this but the change fixed it (ie why is phase in top level of prototxt to begin with if it's always in "include")? If this is a mystery, I can investigate more.